### PR TITLE
[move] add gas metering for SipHash

### PR DESCRIPTION
**Pull request by alinush**
### Description

@vgao1996 requested to fix the missing gas metering.

Also, there are some potential gas metering problems in the native Rust code AFAICT due to the use of the ? operator. @zekun000 and @vgao1996 please take a look at the `TODO`'s in the changes.

Oh, and I tried moving `sip_hash` into `aptos-stdlib` (as per the TODO) but had a naming conflict because both `std` and `aptos_std` are at 0x01, I think.

After fixing that by renaming `aptos_std::hash` to `aptos_std::hashing`, I still had a problem with the `test_genesis_e2e_flow` test, which AFAICT publishes Move modules in the genesis block (I think). So I couldn't move it.

If you tell me how to fix that I can move it to `aptos-stdlib`, which maybe we should move to a new address, to avoid future naming conflicts?

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2801)

### CODE REVIEW

File: aptos-move/framework/src/natives/hash.rs [hunk 1]
```diff
@@ -30,37 +30,40 @@ pub struct SipHashGasParameters {
...
     debug_assert!(args.len() == 1);
 
-    // TODO(Gas): proper gas metering
+    let mut cost = gas_params.base_cost;
 
     let ref_to_val = pop_arg!(args, Reference);
     let arg_type = ty_args.pop().unwrap();
 
     // delegate to the BCS serialization for `Value`
+    // TODO: I'm confused: what cost will this return with the ? operator
```
Comment (zekun000) [inline aptos-move/framework/src/natives/hash.rs:47]: this is an internal error I believe will fail the whole execution (no gas would be charged)

**Review by zekun000 — APPROVED** [review]

File: aptos-move/framework/src/natives/hash.rs (hunk 1 above)
Comment (alinush) [reply aptos-move/framework/src/natives/hash.rs:47]: @vgao1996, this doesn't sound right, I assume? We always want to charge gas, no?

Comment (alinush) [reply aptos-move/framework/src/natives/hash.rs:47]: Anyway, I'm about to change to do the BCS stuff in Move, so these ? operator calls might disappear anyway... Let me see.

Comment (vgao1996) [reply aptos-move/framework/src/natives/hash.rs:47]: Zekun was right on this. This function is supposed to never fail, so if it really happens it's an internal invariant violation. Consider it as "gentler VM crash", where we stop to care about gas metering for the current transaction.

File: aptos-move/framework/move-stdlib/sources/hash.move [hunk 2]
```diff
@@ -3,12 +3,20 @@
 /// The functions in this module are natively declared both in the Move runtime
 /// as in the Move prover's prelude.
 module std::hash {
+    use std::bcs;
+
     // TODO: move sip_hash into aptos_framework
-    native public fun sip_hash<MoveValue>(v: &MoveValue): u64;
-    native public fun sha2_256(data: vector<u8>): vector<u8>;
-    native public fun sha3_256(data: vector<u8>): vector<u8>;
+    public fun sip_hash<MoveValue>(v: &MoveValue): u64 {
```
Comment (vgao1996) [inline aptos-move/framework/move-stdlib/sources/hash.move:9]: Wait a minute, this thing is in our fork of move-stdlib?!

OMG we should definitely move it into aptos-framework ASAP as the comment suggests

File: aptos-move/framework/move-stdlib/sources/hash.move [hunk 3]
```diff
@@ -3,12 +3,20 @@
...
+        let bytes = bcs::to_bytes(v);
+
+        sip_hash_internal(bytes)
+    }
 
     spec sip_hash { // TODO: temporary mockup.
         pragma opaque;
     }
+
+    native fun sip_hash_internal(bytes: vector<u8>): u64;
```
Comment (vgao1996) [inline aptos-move/framework/move-stdlib/sources/hash.move:19]: I would suggest we make this public and name it sip_hash, just to be consistent with other crypto functions.

File: aptos-move/framework/src/natives/hash.rs (hunk 1 above)
Comment (alinush) [reply aptos-move/framework/src/natives/hash.rs]: Okay. Well, it's fixed now after, ahem, ahem, _Moving_ the BCS stuff :)

File: aptos-move/framework/move-stdlib/sources/hash.move (hunk 2 above)
Comment (vgao1996) [inline aptos-move/framework/move-stdlib/sources/hash.move:9]: How about naming it something like `sip_hash_from_value`?

Comment (alinush) [reply aptos-move/framework/move-stdlib/sources/hash.move:9]: I don't feel strongly either way. @zekun000?

File: aptos-move/framework/move-stdlib/sources/hash.move (hunk 3 above)
Comment (alinush) [reply aptos-move/framework/move-stdlib/sources/hash.move:19]: Ah! Okay.

File: aptos-move/framework/move-stdlib/sources/hash.move (hunk 2 above)
Comment (alinush) [reply aptos-move/framework/move-stdlib/sources/hash.move:9]: Yes... I tried to, but can't get it to work unless someone helps me deal with the 0x01 conflict AND pass `test_genesis_e2e_flow`... See PR text above.

Comment (alinush) [reply aptos-move/framework/move-stdlib/sources/hash.move:9]: Okay! Renamed!

Comment (vgao1996) [reply aptos-move/framework/move-stdlib/sources/hash.move]: Yeah looks like there's an on going discussion about this. Let's try to conclude the current PR and move it in another one.

Comment (alinush) [reply aptos-move/framework/move-stdlib/sources/hash.move]: Concluded: moving `sip_hash` from `std::hash` to `aptos_std::aptos_hash` in `aptos-move/framework/aptos-stdlib/sources/hash.move`

Auto-merging once the tests pass!

**Pull request merged**